### DISCUSSION TITLE
The scheduler: a retrospective and charting a new course

### DISCUSSION THREAD

**Discussion by caturria** [AzuraCast/AzuraCast#7603] [General]
Hi @BusterNeece,

I hope you are doing well.

I wanted to create this thread to discuss the past four months where the scheduler is concerned, take responsibility for what went wrong, and discuss how I can best support the project during the holiday season and beyond.

# What I was attempting to solve

I have been using AzuraCast as a hobbyist since 2022. While I won't go into great detail on what my use case is (it's a private passion project involving myself and a close friend rather than a public-facing station), suffice it to say that it frequently involved scheduling large sequences of short tracks, a use case which reveals a number of edge cases with the current scheduling infrastructure.

# Treating the symptoms instead of the root cause

At the end of the day, the issues I was facing (and face again following the rollbacks) revolve around a singular issue: the scheduler does not have access to a source of absolute truth regarding track durations. This is due in part to the rounding of track durations to a single decimal place at processing time, and in part to the scheduler itself internally dealing exclusively in integer seconds.

Since the upcoming song queue is initially built based on inaccurate data and then "corrected" as tracks get closer to their actual playtimes, we run into a number of edge cases that are hard to reproduce and even harder to solve cleanly. Most of the time, playout runs ahead of schedule. The case where a track is dropped because it would end up playing before its scheduled window was the most common manifestation. Worse yet was the case where a scheduled sequence would play its first item (such as a "we'll be right back" message proceeding an ad break), followed by a return to general rotation (because it's too early, but the jingle was already sent to Liquidsoap making its retraction impossible), followed by an eventual return to the scheduled sequence.

Looking back on the work I did to correct these issues during the summer, I now feel that I was implementing patchwork (such as the 'timestamp_scheduled' field, among other things) to address the symptoms rather than trying to make the timings more accurate in the first place. I believe the way to make the scheduler as reliable as possible is to eliminate as much disagreement between anticipated and actual playtimes as humanly possible.

# The 'interrupt other songs' debacle

I'll just come out and say it: I left a mess. For that, I am deeply sorry. The fact of the matter is that I do not personally use the 'interrupt other songs' feature. I personally find interrupted songs jarring outside of very specific, hypothetical circumstances (such as some sort of "breaking news" or emergency cut-in) so I simply never use the feature. The reality, to be completely honest, is that my patches for the issues I was facing inadvertently broke that feature, and I simply did not think to test it. Then my final college term got underway; and even though I receive every academic accommodation I could possibly ask for, getting through my courses as a totally blind person still takes everything out of me. I started to see the issues relating to interrupt piling up, I knew they were going to take more time to address than I had to give at that time, and I genuinely felt bad for largely ignoring a situation I caused for a considerable length of time.

# The current state of affairs

I see that most of my work on the scheduler has been reverted. I harbour no hard feelings, and I appreciate your kind words regarding the reason. In truth, I see it as an opportunity to take a different approach: to deal with the root cause rather than chasing symptoms.

# Moving forward

As promised, I intend to support the development of a new and improved 'interrupt other songs feature' that fabricates shortened track runtimes for tracks which are scheduled to be interrupted, rather than relying on Liquidsoap to provide the functionality and dealing with its limitations. After that, I hope to work with you on making the scheduler "tighter" (reducing the discrepancy between expected and actual playtimes as much as possible; ideally to 1MS or less, but failing that, a tolerance of 100MS or so is workable). I believe that this would address all of those frustrating edge cases without introducing significant complexity into the scheduler itself.

# AutoCue: the elephant in the room

AutoCue is wonderful. But since we rely on Liquidsoap to invoke AutoCue at playtime, we find ourselves in an even worse predicament when it comes to scheduling discrepancies. Ideally we want to know exactly how long a track will play at scheduling time, and notwithstanding truly exceptional circumstances (admin skips track, file is corrupt, etc) we want to stay right on schedule. I propose the following remedies; how do you feel about them:
* When media is processed, storing track durations with three decimal places instead of one
* Representing timestamps in the station_queue table as either microsecond timestamps or seconds with three or four decimal places
* Invoking the AutoCue binary from PHP during the scheduling process if we don't already have cue-in and fade-out times stored in the database, rather than scheduling based on estimates?

I realize this is a lot to take in, and all of this is rather big-picture. I'm prepared to take this as slow as necessary to insure as smooth a transition as possible, but I feel strongly that these measures are necessary in the long run in order to make the scheduler as reliable and edge case free as possible.

I welcome your thoughts on these proposed ideas.

Kind regards,

Caturria.

**Comment by caturria** [comment]
Hi @BusterNeece,

Sorry for the delay; hit a little bit of a roadblock that's kept me from getting back involved in this as quickly as I'd like to.

I fully expect to be able to start devoting time to this again in a few weeks.

Could you please give me your thoughts on this (timing accuracy improvements)?

Also, would it be at all possible at some point for me to have my own branch for this -- like the LS 2.1.X migration had -- so that it can be voluntarily opted into via the installer for testing?

Kind regards,

Caturria.

**Reply by BusterNeece** [reply]
@caturria I am fully in support of the idea of improving the granularity of our timing from the full-second mark in most places to a decimal or floating point value. There are pros and cons to either approach: the decimal would be processed as a string, but we'd have the assurance of more accuracy with it (we have the `BigDecimal` library for passing around decimal values already, so some of the complexity there is reduced), but the floating point would always cast as a `float`, it would just have the potential for some rounding errors.

Also, yes, we can definitely do a 'dev' branch for this that users can test via our Docker published builds. If you submit a pull request for the changes, I'll do the necessary work under the hood to give you a destination branch to play with.

**Reply by BusterNeece** [reply]
@caturria Thinking about this a little more...both modern PHP and modern MariaDB/MySQL will handle datetimes down to the microsecond resolution. If we say that duration is a millisecond value on media tracks, we can pass that around as a whole integer value (since we're only on 64-bit OSes, INT_MAX has enough overhead for this) and then use UTC DateTime values with millisecond precision for everything else in the chain.

**Comment by caturria** [comment]
Absolutely.

This is only half the battle though as AutoCue being handled by Liquidsoap is the most significant source of disagreement between expected and actual play time.

How do you feel about the long-term goal of invoking AutoCue from the queue worker at scheduling time instead of at play time?

The scheduler needs access to a source of absolute truth in order to make appropriate decisions, so that's the major benefit. Secondary benefits include adjusting AutoCue settings without requiring a station restart and going back to just one mandatory Liquidsoap prefetch instead of two or more.

**Reply by BusterNeece** [reply]
@caturria I mean, in the end I absolutely want to get there. Doing so may mean just manually invoking a little Liquidsoap script whose whole job is to run AutoCue, calculate the values and then fire off an API call back to us to alert us when it's done.
